### ISSUE TITLE
[platform-views] Deadlock caused by thread merging

### ISSUE THREAD

**Issue by 0xZOne** [flutter/flutter#94524]
As shown below, If thread merging occurs before the task in the red box is posted, this results in a deadlock as the task will never be posted to the raster thread that is blocked on a latch.

<img width="731" alt="截屏2021-12-02 下午2 11 00" src="https://user-images.githubusercontent.com/26625149/144368322-e5f0f947-d543-4452-baf7-22d193bc3c55.png">

Related issue: #80491

<details>
  <summary>Crash logs</summary>

```
#0  0x00000001bd675550 in __psynch_cvwait ()
#1  0x00000001ddaa396c in _pthread_cond_wait$VARIANT$armv81 ()
#2  0x000000019b22b2b8 in std::__1::condition_variable::wait(std::__1::unique_lock<std::__1::mutex>&) ()
#3  0x00000001164745c0 in fml::ManualResetWaitableEvent::Wait()
#4  0x0000000117552000 in flutter::PlatformView::NotifyCreated()
#5  0x000000011639ddc4 in -[FlutterViewController surfaceUpdated:]
#6  0x0000000104e98230 in -[FLBFlutterViewContainer surfaceUpdated:]
#7  0x0000000104e98698 in -[FLBFlutterViewContainer viewDidAppear:]
#8  0x00000001072deb90 in -[LTaoMiniDetailFlutterViewController viewDidAppear:] ()
#9  0x0000000186598c4c in -[UIViewController _setViewAppearState:isAnimating:] ()
#10  0x0000000186680570 in -[UIViewController __viewDidAppear:] ()
#11  0x0000000186606420 in -[UIViewController _endAppearanceTransition:] ()
#12  0x0000000186792a5c in -[UINavigationController navigationTransitionView:didEndTransition:fromView:toView:] ()
#13  0x0000000105272ad4 in -[UINavigationController(AliHA) aliHASwizzledNavigationTransitionView:didEndTransition:fromView:toView:]
#14  0x0000000186845fe8 in __49-[UINavigationController _startCustomTransition:]_block_invoke ()
#15  0x000000018662ac34 in -[_UIViewControllerTransitionContext completeTransition:] ()
#16  0x0000000186d55468 in __53-[_UINavigationParallaxTransition animateTransition:]_block_invoke.158 ()
#17  0x00000001866ab000 in -[UIViewAnimationBlockDelegate _didEndBlockAnimation:finished:context:] ()
#18  0x000000018657d9a8 in -[UIViewAnimationState sendDelegateAnimationDidStop:finished:] ()
#19  0x0000000186591210 in -[UIViewAnimationState animationDidStop:finished:] ()
#20  0x0000000186591364 in -[UIViewAnimationState animationDidStop:finished:] ()
#21  0x0000000187c66db0 in CA::Layer::run_animation_callbacks(void*) ()
#22  0x000000010f62e1c8 in _dispatch_client_callout ()
#23  0x000000010f63c8d8 in _dispatch_main_queue_callback_4CF ()
#24  0x00000001840e57f8 in __CFRUNLOOP_IS_SERVICING_THE_MAIN_DISPATCH_QUEUE__ ()
#25  0x00000001840a30f8 in __CFRunLoopRun ()
#26  0x00000001840b5d8c in CFRunLoopRunSpecific ()
#27  0x000000019e1c29a0 in GSEventRunModal ()
#28  0x00000001868eb018 in -[UIApplication _run] ()
#29  0x000000018667f29c in UIApplicationMain ()
#30  0x0000000104d9919c in main
#31  0x000000010f5a4190 in start ()
```

```
```

```
```

</details>

**Comment by maheshmnj** [comment]
Hi @0xZOne, Thanks for filing the issue. Is this potential deadlock from the latest engine revision? Can you please share the output of `flutter doctor -v`?

Thanks

**Comment by 0xZOne** [comment]
> Hi @0xZOne, Thanks for filing the issue. Is this potential deadlock from the latest engine revision? Can you please share the output of `flutter doctor -v`?
> 
> Thanks

Yes, I think so.

**Comment by maheshmnj** [comment]
@0xZOne, looking at your flutter doctor it looks like you are using an older version of flutter https://github.com/flutter/flutter/releases/tag/1.22.6, Please upgrade to latest `stable 2.5.3` and see if you are still experiencing the deadlock. 
if the issue persists feel free to write back with the updated output of `flutter doctor -v` along with reproducible steps.

Thanks

**Comment by 0xZOne** [comment]
> @0xZOne, looking at your flutter doctor it looks like you are using an older version of flutter https://github.com/flutter/flutter/releases/tag/1.22.6, Please upgrade to latest `stable 2.5.3` and see if you are still experiencing the deadlock. if the issue persists feel free to write back with the updated output of `flutter doctor -v` along with reproducible steps.
> 
> Thanks

Yes, We found the deadlock in our actual ios app that uses flutter 1.22.6, as the log show. The deadlock only occurs in specific scenes occasionally and is hard to reproduce.  Maybe we will not upgrade to 2.5.x soon.

But, According to the code logic, I believe that the deadlock still exists in the latest version.

**Comment by maheshmnj** [comment]
Thanks for the info, Leaving this issue open for further insights from the team.

**Comment by jmagman** [comment]
https://github.com/flutter/engine/blob/d8a3b066e07522ee976ae7d8cb1af458f99a4ff1/shell/common/platform_view.cc#L69

**Comment by 0xZOne** [comment]
> Can you share complete crash logs to help debug the issue?

Complete crash logs have been added.

**Comment by maheshmnj** [comment]
@flutter-symbolizer-bot [#93791 (comment)](https://github.com/flutter/flutter/issues/94524#issue-1069187045) engine#2f0af37152 ios release

**Comment by maheshmnj** [comment]
@flutter-symbolizer-bot [#93791 (comment)](https://github.com/flutter/flutter/issues/94524#issue-1069187045) engine#2f0af37152 ios release x86

**Comment by 0xZOne** [comment]
@maheshmnj

The key symbol stack is already there.

```
#0  0x00000001bd675550 in __psynch_cvwait ()
#1  0x00000001ddaa396c in _pthread_cond_wait$VARIANT$armv81 ()
#2  0x000000019b22b2b8 in std::__1::condition_variable::wait(std::__1::unique_lock<std::__1::mutex>&) ()
#3  0x00000001164745c0 in fml::ManualResetWaitableEvent::Wait()
#4  0x0000000117552000 in flutter::PlatformView::NotifyCreated()
#5  0x000000011639ddc4 in -[FlutterViewController surfaceUpdated:]
#6  0x0000000104e98230 in -[FLBFlutterViewContainer surfaceUpdated:]
#7  0x0000000104e98698 in -[FLBFlutterViewContainer viewDidAppear:]
#8  0x00000001072deb90 in -[LTaoMiniDetailFlutterViewController viewDidAppear:] ()
#9  0x0000000186598c4c in -[UIViewController _setViewAppearState:isAnimating:] ()
#10  0x0000000186680570 in -[UIViewController __viewDidAppear:] ()
#11  0x0000000186606420 in -[UIViewController _endAppearanceTransition:] ()
#12  0x0000000186792a5c in -[UINavigationController navigationTransitionView:didEndTransition:fromView:toView:] ()
#13  0x0000000105272ad4 in -[UINavigationController(AliHA) aliHASwizzledNavigationTransitionView:didEndTransition:fromView:toView:]
#14  0x0000000186845fe8 in __49-[UINavigationController _startCustomTransition:]_block_invoke ()
#15  0x000000018662ac34 in -[_UIViewControllerTransitionContext completeTransition:] ()
#16  0x0000000186d55468 in __53-[_UINavigationParallaxTransition animateTransition:]_block_invoke.158 ()
#17  0x00000001866ab000 in -[UIViewAnimationBlockDelegate _didEndBlockAnimation:finished:context:] ()
#18  0x000000018657d9a8 in -[UIViewAnimationState sendDelegateAnimationDidStop:finished:] ()
#19  0x0000000186591210 in -[UIViewAnimationState animationDidStop:finished:] ()
#20  0x0000000186591364 in -[UIViewAnimationState animationDidStop:finished:] ()
#21  0x0000000187c66db0 in CA::Layer::run_animation_callbacks(void*) ()
#22  0x000000010f62e1c8 in _dispatch_client_callout ()
#23  0x000000010f63c8d8 in _dispatch_main_queue_callback_4CF ()
#24  0x00000001840e57f8 in __CFRUNLOOP_IS_SERVICING_THE_MAIN_DISPATCH_QUEUE__ ()
#25  0x00000001840a30f8 in __CFRunLoopRun ()
#26  0x00000001840b5d8c in CFRunLoopRunSpecific ()
#27  0x000000019e1c29a0 in GSEventRunModal ()
#28  0x00000001868eb018 in -[UIApplication _run] ()
#29  0x000000018667f29c in UIApplicationMain ()
#30  0x0000000104d9919c in main
#31  0x000000010f5a4190 in start ()

```

**Comment by dnfield** [comment]
Are you sure the merger is happening _before_ the red box?

The only way I could see this deadlocking is if the threads get merged _immediately_ after calling `RunNowOrPostTask`. Perhaps this is a general bug, and we have to flush the task queues before merging threads?

**Comment by dnfield** [comment]
I suspect there might be a bug in the thread merger code - I'm not sure if I'mr eading it right or not though. It seems like before merging threads, we should make sure to flush their task queues, otherwise we could run into these deadlocks in other places. @iskakaushik might know more about this, or @cyanglaz

**Comment by dnfield** [comment]
Chinmay tells me that we definitely don't flush the task queue(s) before merging.

**Comment by dnfield** [comment]
Duplicate of https://github.com/flutter/flutter/issues/94524?

**Comment by dnfield** [comment]
Sorry, I mean tto ask if it's a duplicate of https://github.com/flutter/flutter/issues/80491

**Comment by 0xZOne** [comment]
> Are you sure the merger is happening _before_ the red box?
> 
> The only way I could see this deadlocking is if the threads get merged _immediately_ after calling `RunNowOrPostTask`. Perhaps this is a general bug, and we have to flush the task queues before merging threads?

@dnfield   Sorry, perhaps I didn't make myself clear (I updated the issue description).  

I consider that thread merging and platform thread posting sync task to raster require a synchronization mechanism.

**Comment by 0xZOne** [comment]
> @0xZOne - I agree there's a synchronization issue, but we should fix this in the general case for thread merging, rather than for one particular broken use case :)

Yes, I totally agree with you.

**Comment by cyanglaz** [comment]
Unassigned myself as I don't have time to work on this at this moment.

**Comment by wamynobe** [comment]
Did this [PR](https://github.com/flutter/engine/pull/53826) fix this issue?
